demo/src/App.js

```diff
@@ -1,48 +1,117 @@
-  constructor() {
-    super();
+const DEFAULT_PATH = "/semantic-role-labeling"
+
+// The App is just a react-router wrapped around the Demo component.
+const App = () => (
+  <Router>
+    <div>
+      <Route exact path="/" render={() => (
+        <Redirect to={DEFAULT_PATH}/>
+      )}/>
+      <Route path="/:model/:slug?" component={Demo}/>
+    </div>
+  </Router>
+)
+
+class Demo extends React.Component {
+  constructor(props) {
+    super(props);
+
+    // React router supplies us with a model name and (possibly) a slug.
+    const { model, slug } = props.match.params;
+
     this.state = {
-      selectedModel: "srl", // valid values: "srl", "mc", "te"
-      rawOutput: "",
+      slug: slug,
+      selectedModel: model,
+      requestData: null,
+      responseData: null
     };
-    this.changeModel = this.changeModel.bind(this);
-  }
 
-  changeModel(thisModel) {
-    this.setState({
-      selectedModel: thisModel,
+    // We'll need to pass this to the Header component so that it can clear
+    // out the data when you switch from one model to another.
+    this.clearData = () => {
+      this.setState({requestData: null, responseData: null})
+    }
+
+    // Our components will be using history.push to change the location,
+    // and they will be attaching any `requestData` and `responseData` updates
+    // to the location object. That means we need to listen for location changes
+    // and update our state accordingly.
+    props.history.listen((location, action) => {
+      const { state } = location;
+      if (state) {
+        const { requestData, responseData } = state;
+        this.setState({requestData: requestData, responseData: responseData})
+      }
     });
   }
 
-  render() {
+  // We also need to update the state whenever we receive new props from React router.
+  componentWillReceiveProps({ match }) {
+    const { model, slug } = match.params;
+    this.setState({selectedModel: model, slug: slug});
+  }
 
-    const { selectedModel } = this.state;
+  // After the component mounts, we check if we need to fetch the data
+  // for a permalink.
+  componentDidMount() {
+    const { slug, responseData } = this.state;
+
+    // If this is a permalink and we don't yet have the data for it...
+    if (slug && !responseData) {
+      // Make an ajax call to get the permadata,
+      // and then use it to update the state.
+      fetch('http://localhost:8000/permadata', {
+        method: 'POST',
+        headers: {
+          'Accept': 'application/json',
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({"slug": slug})
+      }).then(function(response) {
+        return response.json();
+      }).then((json) => {
+        const { requestData, responseData } = json;
+        this.setState({requestData: requestData, responseData: responseData});
+      }).catch((error) => {
+        this.setState({outputState: "error"});
+        throw error; // todo(michaels): is this right?
```

Comment:
Since you've got a UI state conveying the error, I would just do `console.error(error)`.